### PR TITLE
Add constant dynamic unit tests

### PR DESCRIPTION
**Pull request by r30shah**
Add Java 11 tests for Constant Dynamic mainly involves String object, NULL object and Primitive data types objects. Variation covers running this test in interpreted mode as well as specific JIT compilation modes which checks out JIT's handling of resolved and unresolved Constant Dynamic cases.

Signed-off-by: Rahil Shah <rahil@ca.ibm.com>

### CODE REVIEW

**Comment by r30shah** [comment]
@yanluo7 @fjeremic  These are the tests for Constant Dynamic which includes Testing following primitive data types 
1. Int
2. Float
3. Double
4. Long 
It also tests Null and String Objects.
@smlambert  Can I get a review from Testing team as well? Especially for [second commit](https://github.com/eclipse/openj9/pull/2870/commits/3e1c151c4daa4ae2a632435a56830eccaff5dfff)
ASM generator that we usually got is from release branch which did not have support for Constant Dynamic. So I changed to to get from Snapshots which I am not sure is fine or not.

File: test/functional/Java11andUp/playlist.xml [hunk 1]
```diff
@@ -89,6 +89,45 @@
...
+		-->
+	<test>
+		<testCaseName>CondyPrimitive</testCaseName>
+		<variations>
+			<variation>'-Xint'</variation>
+			<variation>'-Xjit:count=1,disableAsyncCompilation'</variation>
+			<variation>'-Xjit:count=0,disableAsyncCompilation'</variation>
+		</variations>
+		<command>$(JAVA_COMMAND) $(JVM_OPTIONS) \
+			-Xverify:none \
```
Comment (DanHeidinga) [inline test/functional/Java11andUp/playlist.xml]: This is a temporary option until the verifier changes for condy have been merged...  which I thought had already occurred.  Can you confirm this is still required with the latest in `master`?

File: test/functional/Java11andUp/src/org/openj9/test/condy/BootstrapMethods.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+public class BootstrapMethods {
+
+
+public static Object bootstrap_constant_string(MethodHandles.Lookup l, String name, Class<?> c, String s) {
+	return s;
+}
+
+public static int bootstrap_constant_int(MethodHandles.Lookup l, String name, Class<?> c, int v) {
+	return v;
+	}
```
Comment (DanHeidinga) [inline test/functional/Java11andUp/src/org/openj9/test/condy/BootstrapMethods.java]: Formatting - the methods should be indented like:

```java
public class BootstrapMethods {
	public static Object bootstrap_constant_string(MethodHandles.Lookup l, String name, Class<?> c, String s) {
		return s;
	}
}

```

File: test/functional/Java11andUp/src/org/openj9/test/condy/PrimitiveCondyTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import org.testng.annotations.Test;
+import org.testng.log4testng.Logger;
+
+public class PrimitiveCondyTest {
+	 
+	@Test(groups = { "level.sanity" })
+	public void testCondyNull() {
+		Assert.assertNull(PrimitiveCondyMethods.condy_return_null());
+		Assert.assertNull(PrimitiveCondyMethods.condy_return_null());
+		Assert.assertNull(PrimitiveCondyMethods.condy_return_null());
```
Comment (DanHeidinga) [inline test/functional/Java11andUp/src/org/openj9/test/condy/PrimitiveCondyTest.java]: Can you add a comment explaining why this needs to be invoked 3 times?

File: test/TestConfig/scripts/tools/getDependencies.pl [hunk 4]
```diff
@@ -66,9 +66,9 @@
 # Define a a hash for each dependent jar
 # Contents in the hash should be: url => , fname =>, sha1 =>
 my %asm_all = (
-	url => 'http://central.maven.org/maven2/org/ow2/asm/asm-all/6.0_BETA/asm-all-6.0_BETA.jar',
+	url => 'https://repository.ow2.org/nexus/content/repositories/snapshots/org/ow2/asm/asm/7.0-beta-SNAPSHOT/asm-7.0-beta-20180911.182645-11.jar',
```
Comment (fjeremic) [inline test/TestConfig/scripts/tools/getDependencies.pl]: @smlambert @llxia we had to upgrade ASM to a newer level for the condy bytecode support. I'm just verifying this change is ok from the test perspective. Please let us know. Thanks!

Comment (llxia) [reply test/TestConfig/scripts/tools/getDependencies.pl]: LGTM. 
- https://ci.eclipse.org/openj9/job/test.getDependency/ - may need to manually launch after  merge. It is schedule to run every hour. If any test build is launched between the merge and new test.getDependency build, the test build will fail. 
- Same for https://ci.adoptopenjdk.net/view/all/job/test.getDependency/ and internal Jenkins

File: test/functional/Java11andUp/playlist.xml [hunk 5]
```diff
@@ -89,6 +89,45 @@
...
+			This variation runs test under -Xjit:count=1 option which makes sure that ConstantDynamic is resoleved before we compile a method
+			to test JIT's handling of resolved constant dynamic
+		Variation 3:
+			This variation runs test under -Xjit:count=0 option means JIT faces unresolved constant dynamic and it needs to resolve them before
+			using it.
+		-->
+	<test>
+		<testCaseName>CondyPrimitive</testCaseName>
+		<variations>
+			<variation>'-Xint'</variation>
```
[the remark this answers is not in the document]
Comment (r30shah) [reply test/functional/Java11andUp/playlist.xml]: Changed to `NoOptions` in  https://github.com/eclipse/openj9/pull/2870/commits/2d8f421c55cf7ae39bb68c90bac24b683061fe29

File: test/functional/Java11andUp/playlist.xml [hunk 6]
```diff
@@ -89,6 +89,45 @@
...
+			<variation>'-Xjit:count=1,disableAsyncCompilation'</variation>
+			<variation>'-Xjit:count=0,disableAsyncCompilation'</variation>
```
[the remark this answers is not in the document]
Comment (r30shah) [reply test/functional/Java11andUp/playlist.xml]: Updated to `-Xjit:count=1,disableAsyncCompilation,rtResolve` in https://github.com/eclipse/openj9/pull/2870/commits/2d8f421c55cf7ae39bb68c90bac24b683061fe29

File: test/functional/Java11andUp/playlist.xml (hunk 1 above)
Comment (r30shah) [reply test/functional/Java11andUp/playlist.xml]: Verified with master repo. We do not need it. Removed in https://github.com/eclipse/openj9/pull/2870/commits/2d8f421c55cf7ae39bb68c90bac24b683061fe29

File: test/functional/Java11andUp/src/org/openj9/test/condy/PrimitiveCondyTest.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ *******************************************************************************/
+package org.openj9.test.condy;
+
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import org.testng.log4testng.Logger;
+
+public class PrimitiveCondyTest {
+	/**
+	 * In all of the following test methods, methods from PrimitiveCondyMethods are called 3 times
```
Comment (r30shah) [inline test/functional/Java11andUp/src/org/openj9/test/condy/PrimitiveCondyTest.java]: @DanHeidinga  I have commented in details why I chose methods to call multiple times (3).
Basically for testing JIT with resolved cases I need to make sure it is ran atleast one time in interpreted mode. Also
for unresolved cases (count=0), first invocation of the method will run into path where we are calling Bootstrapo methods which resolves condy and branch to call helper to resolve this is patched to NOP. 
Second invocation makes sure that once Condy is resolved by JIT helper, JIT loads correct value.

**Comment by fjeremic** [comment]
I'll wait until @DanHeidinga gives an approval before launching builds for this just in case you may have additional comments on the change.

**Comment by r30shah** [comment]
Let me squash the commits. I will keep grabbing latest ASM change in separate PR.

**Comment by fjeremic** [comment]
Jenkins test sanity xlinux,plinux,zlinux JDK8,JDK11

**Comment by r30shah** [comment]
It seems like we need asm-all (Failures corresponds to misssing components like `asm/tree/*`
Which are present in released version of `asm-all.jar` which seems to combine all components of asm in single jar. 
Trying to find out similar `asm-all.jar` for 7.0 snapshots.

**Comment by r30shah** [comment]
@fjeremic  @llxia I made a change, specifically with getting asm commit. Now we will have two asm jars, 
One `asm-all.jar` required by all tests and another `asm-7.0.jar` needed for Java 11 tests. This is temporary change , as expecting to have all combined jar available for 7.0 version once released. Made corresponding changes Java11andUp/build.xml as well.

**Comment by fjeremic** [comment]
JDK11 failures due to #2708. This was way before we merged any of the JIT changes in this area so it cannot be related to this change.
